samza-sql/src/main/java/org/apache/samza/sql/translator/JoinTranslator.java

```
@@ -0,0 +1,319 @@
+ * in the join condition.
+ */
+public class JoinTranslator {
+
+  private static final Logger log = LoggerFactory.getLogger(JoinTranslator.class);
+  private int joinId;
+  private SourceResolver sourceResolver;
+
+  public JoinTranslator(int joinId, SourceResolver sourceResolver) {
+    this.joinId = joinId;
+    this.sourceResolver = sourceResolver;
+  }
+
+  public void translate(final LogicalJoin join, final TranslatorContext context) {
+
+    // Do the validation of join query
+    validateJoinQuery(join);
+
+    boolean isTablePosOnRight = isTable(join.getRight());
+    List<Integer> streamIds = new LinkedList<>();
+    List<Integer> tableIds = new LinkedList<>();
+
+    // Get the stream and table indices corresponding to the fields given in the join condition.
+    getStreamAndTableKeyIds(((RexCall) join.getCondition()).getOperands(), join, isTablePosOnRight, streamIds, tableIds);
+
+    MessageStream<SamzaSqlRelMessage> inputTable =
+        isTablePosOnRight ?
+            context.getMessageStream(join.getRight().getId()) : context.getMessageStream(join.getLeft().getId());
+
+    SamzaSqlCompositeKeySerde keySerde =
+        (SamzaSqlCompositeKeySerde) new SamzaSqlCompositeKeySerdeFactory().getSerde(null, null);
+    SamzaSqlRelMessageSerde relMsgSerde =
+        (SamzaSqlRelMessageSerde) new SamzaSqlRelMessageSerdeFactory().getSerde(null, null);
+
+    // Create a table backed by RocksDb store with the fields in the join condition as composite key and relational
+    // message as the value. Send the messages from the input stream denoted as 'table' to the created table store.
+    Table<KV<SamzaSqlCompositeKey, SamzaSqlRelMessage>> table =
+        context.getStreamGraph()
+            .getTable(new RocksDbTableDescriptor("table_" + joinId)
+                .withSerde(KVSerde.of(keySerde, relMsgSerde)));
+
+    inputTable
+        .map(m -> new KV(createSamzaSqlCompositeKey(m, tableIds), m))
+        .sendTo(table);
```

can you throw a error message, or increase a counter. If the hash of the composite key does not match the hash assigned to this task (breaking one of the assumption in the class comment)